symengine/expression.h

```diff
@@ -0,0 +1,108 @@
+/**
+ *  \file expression.h
+ *  Includes the expression class
+ *
+ **/
+
+#ifndef SYMENGINE_EXPRESSION_H
+#define SYMENGINE_EXPRESSION_H
+
+#include <iostream>
+#include <type_traits>
+
+#include <symengine/add.h>
+#include <symengine/basic.h>
+#include <symengine/symengine_rcp.h>
+#include <symengine/integer.h>
+#include <symengine/mul.h>
+#include <symengine/pow.h>
+#include <symengine/symbol.h>
+
+namespace SymEngine
+{
+
+class expression
+{
+	public:
+		expression():m_basic(new Integer(0)) {}
+		template <typename T, typename = typename std::enable_if<std::is_constructible<RCP<const Basic>,T &&>::value>::type>
+		expression(T &&o):m_basic(std::forward<T>(o)) {}
+		expression(int n):m_basic(new Integer(n)) {}
+		expression(const expression &) = default;
+		expression(expression &&other) noexcept : m_basic(std::move(other.m_basic)) {}
+		expression &operator=(const expression &) = default;
+		expression &operator=(expression &&other) noexcept
+		{
+			if (this != &other) {
+				*this = std::move(other);
+			}
+			return *this;
+		}
+		~expression() noexcept {}
+		friend std::ostream &operator<<(std::ostream &os, const expression &expr)
+		{
+			os << (*expr.m_basic.get());
+			return os;
+		}
+		friend expression operator+(const expression &a, const expression &b)
```

Comment:
It's allowed for this class, since it is a value class (behaving like a
value).

Sent from my mobile phone.
On Jul 3, 2015 11:13 AM, "Sumith" <notifications@github.com> wrote:

> In symengine/expression.h
> <https://github.com/sympy/symengine/pull/500#discussion_r33874663>:
>
> > +		expression(expression &&other) noexcept : m_basic(std::move(other.m_basic)) {}
> > +		expression &operator=(const expression &) = default;
> > +		expression &operator=(expression &&other) noexcept
> > +		{
> > +			if (this != &other) {
> > +				*this = std::move(other);
> > +			}
> > +			return *this;
> > +		}
> > +		~expression() noexcept {}
> > +		friend std::ostream &operator<<(std::ostream &os, const expression &expr)
> > +		{
> > +			os << (*expr.m_basic.get());
> > +			return os;
> > +		}
> > +		friend expression operator+(const expression &a, const expression &b)
>
> @certik <https://github.com/certik> We never overload operators, do we?
> Or is it allowed for this class?
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/sympy/symengine/pull/500/files#r33874663>.
>